### PR TITLE
Data curation with configure parameter `query_group`

### LINKED ISSUE #237: [MCLOUD] Allow for data curation
For Materials Cloud servers it would be beneficial to only choose a curated number of data from an AiiDA database.

One solution for this, as suggested by @giovannipizzi, is to group these `Node`s in a `Group` and only expose this `Group`'s `Node`s.
Another solution could be to check for a specific `Node.extra` and only expose a `Node` on a condition of this `Node.extra`.

In order to determine whether this feature should be activated or not a special configuration setting might be introduced.

### PR DESCRIPTION
**Pull request by unkcpz**
Fix #237

By setting `query_group` in config will ask optimade to only query the node in the corresponding group. 
The setting will also influence the Mongo database creation.

- [x] New archive with group to test this new feature, where I append 5 more structure data. If the `query_group` is not set in the config one of the test failed `FAILED tests/server/routers/test_structures.py::TestStructuresEndpoint::test_structures_endpoint_data - assert 1625 == 1620`

### CODE REVIEW

**Comment by unkcpz** [comment]
I still need to do a test to see whether `init` and `calc` works well with this feature after the migration to 2.0 PR merged. 
If this is working, I think this can also solve the problem of the lithium battery database optimade field initialization always fails with timeout in container.

**Comment by unkcpz** [comment]
I rebase this PR on `update/aiida-v2` so @eimrek can test this on the materials cloud.

**Comment by unkcpz** [comment]
The `optimade-sample` profile already have a group ‘all_cifdata’ which consist of all cifdata node. To test data curated feature, checkout to this aiida-optimade branch, and in optimade config file, add an parameter ‘query_group’ for the group that you need to query for. See the example file at https://github.com/aiidateam/aiida-optimade/blob/98800bea3500f21fd762635e0ced7171b649315e/tests/static/test_data_curation_config.json

Things need to test:
- [ ] Running `aiida-optimade -p optimade-sample init`, it should only update nodes on the group but not other nodes(StructureData node in this case).
- [ ] Make a query ‘.../v1/structures’ on that database and only 528 results returned. I only test this locally but it should work. 
- [ ] A stress test on MC3D if the above tests are fine. I think we can create a group called ‘final_structures’ and import mc3d-structure archive. Since all these nodes are from mc3d, so it should not add new structure data into mc3d profile.

**Comment by eimrek** [comment]
Hi @unkcpz,

I ran some tests on the materials cloud aiida server.

I enabled the 'all_cifdata' `query_group` for optimade-sample and now indeed 528 structures are returned by optimade (see https://dev-aiida.materialscloud.org/optimade-sample/optimade/v1/structures).

Then, to benchmark on mc3d, I prepared the following script:

```
import requests
import json

with open("./mc3d_data.json", 'r') as fh:
    discover_json = json.loads(fh.read())

uuid_list = []

for label in discover_json:
    for structure in discover_json[label]:
        uuid = structure['uuid_structure']
        uuid_list.append(uuid)

uri = "http://localhost:3757/v1/structures"
for uuid in uuid_list[:1000]:
    response = requests.get(uri+f"?filter=immutable_id=%22{uuid}%22")
    assert response.status_code == 200
    assert response.json()['meta']['data_returned'] == 1
```

where `mc3d_data.json` is the file that contains the uuids used on the MC3D discover section.

Basically, I access the first 1000 uuids served on discover via optimade.

Then, as you suggested, I set up MC3D profile and imported mc3d-structures as a `final_structures` group on the dev-aiida-2 machine.

And I timed the aforementioned script for three cases:

1. The production machine, still running a separate `mc3d-structures` profile:

```
(base) ubuntu@aiida:~/optimade-perf-test$ time python benchmark_optimade.py 

real	1m36.336s
user	0m5.644s
sys	0m0.484s
```

2. On dev-aiida-2 machine, setting `'query_group': null` in the optimade docker container and using the full mc3d profile:

```
(base) ubuntu@dev-aiida-2:~/optimade-perf-test$ time python benchmark_optimade.py 

real	1m10.057s
user	0m4.544s
sys	0m0.505s
```

3. On dev-aiida-2 machine, setting `'query_group':  'final_structures'` in the optimade docker container and using the full mc3d profile:

```
(base) ubuntu@dev-aiida-2:~/optimade-perf-test$ time python benchmark_optimade.py 

real	1m24.529s
user	0m4.710s
sys	0m0.525s
```

Ran everything 2-3 times and the times seem to be consistent.

But do I understand correctly, this was not supposed to make it more performant, but just to expose only the relevant structures, right?

In this case, everything seems to work and the performance is equivalent to before.

**Comment by unkcpz** [comment]
@eimrek thanks! It is really a great news. 🎉

> But do I understand correctly, this was not supposed to make it more performant, but just to expose only the relevant structures, right?

Yes, it is not targeting the performance. But I am a bit surprised to see query on full-MC3D and on MC3D-structure have the same speed.
But it is all good to go I think. But I think @CasperWA better to have a final check. (But this PR is based on the migrate-v2.0 https://github.com/aiidateam/aiida-optimade/pull/331 so @eimrek can do the test. It will be much clean if that one is merged and I rebase this PR.)

BTW. I am looking forward to seeing if this feature helps the li-battery profile available.

**Comment by unkcpz** [comment]
I rebase this PR, and since @eimrek already uses it, you can have a final check.

File: .github/workflows/ci_tests.yml [hunk 1]
```diff
@@ -103,6 +103,14 @@ jobs:
         OPTIMADE_DATABASE_BACKEND: mongodb
       run: pytest -v --cov=./aiida_optimade/ --cov-report=xml:mongo_cov.xml --durations=20
 
+    - name: Test with pytest (Data Curation)
+      env:
+        AIIDA_TEST_BACKEND: ${{ matrix.backend }}
+        AIIDA_TEST_PROFILE: test_${{ matrix.backend }}
+        AIIDA_PROFILE: test_${{ matrix.backend }}
+        PYTEST_OPTIMADE_CONFIG_FILE: ./tests/static/test_data_curation_config.json
+      run: pytest -v --cov=./aiida_optimade/ --cov-report=xml --durations=20 tests/server/routers/test_structures.py::test_structures_endpoint_data
```
Comment (CasperWA) [inline .github/workflows/ci_tests.yml]: Ensure the current coverage report is not overwritten (the one created in the "Test with pytest (AiiDA)" step).
```suggestion
      run: pytest -v --cov=./aiida_optimade/ --cov-report=xml --cov-append --durations=20 tests/server/routers/test_structures.py::test_structures_endpoint_data
```

File: README.md [hunk 2]
```diff
@@ -153,6 +153,12 @@ You can configure the server with the `aiida_optimade/config.json` file or set c
 
 To learn more about this, see the [`optimade-python-tools`](https://github.com/Materials-Consortia/optimade-python-tools) repository.
 
+### Using AiiDA group for data curated
```
Comment (CasperWA) [inline README.md]:
```suggestion
### Using AiiDA group for curated data
```

File: README.md [hunk 3]
```diff
@@ -153,6 +153,12 @@ You can configure the server with the `aiida_optimade/config.json` file or set c
...
+
+The [AiiDA group](https://aiida.readthedocs.io/projects/aiida-core/en/latest/howto/data.html#organizing-data) can be used to curated data and server in bias.
```
Comment (CasperWA) [inline README.md]: I don't understand what you mean by "and server in bias".
```suggestion
An [AiiDA Group](https://aiida.readthedocs.io/projects/aiida-core/en/latest/howto/data.html#organizing-data) can be used to curate data and serve only this curated data through the OPTIMADE server. 
```

File: README.md [hunk 4]
```diff
@@ -153,6 +153,12 @@ You can configure the server with the `aiida_optimade/config.json` file or set c
 
 To learn more about this, see the [`optimade-python-tools`](https://github.com/Materials-Consortia/optimade-python-tools) repository.
 
+### Using AiiDA group for data curated
+
+The [AiiDA group](https://aiida.readthedocs.io/projects/aiida-core/en/latest/howto/data.html#organizing-data) can be used to curated data and server in bias. 
+By setting the `query_group` in `config.json` can only server the structure data from the group.
```
Comment (CasperWA) [inline README.md]:
```suggestion
Setting the `query_group` option in `config.json` will ensure only the valid (`StructureData`, `CifData`) data nodes in the given AiiDA Group will be served.
```

File: README.md [hunk 5]
```diff
@@ -153,6 +153,12 @@ You can configure the server with the `aiida_optimade/config.json` file or set c
...
+Set the `query_group` parameter to `null` (default) to server all structure data from the database.
```
Comment (CasperWA) [inline README.md]:
```suggestion
Set the `query_group` parameter to `null` (default) to serve all structure data from the database.
```

File: aiida_optimade/config.py [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+
+from pydantic import Field
+
+from optimade.server.config import ServerConfig
+
+
+class CustomServerConfig(ServerConfig):
+
+    query_group: Optional[str] = Field(
+        None, description="The aiida group where curate the data for query."
```
Comment (CasperWA) [inline aiida_optimade/config.py]:
```suggestion
        None, description="The AiiDA Group containing the data that will be served, allowing one to serve a curated set of data from a given database."
```

File: aiida_optimade/routers/structures.py [hunk 7]
```diff
@@ -19,11 +19,12 @@
 from aiida_optimade.mappers import StructureMapper
 from aiida_optimade.models import StructureResource
 from aiida_optimade.routers.utils import get_entries, get_single_entry, close_session
-
+from aiida_optimade.config import CONFIG
```
Comment (CasperWA) [inline aiida_optimade/routers/structures.py]: I would move this up to be the first import line from `aiida_optimade` - just to keep the import lines ordered alphabetically.
This is a formatting thing, not a requirement.

File: tests/static/test_mongo_config.json [hunk 8]
```diff
@@ -1,5 +1,6 @@
 {
     "debug": true,
+    "query_group": null,
```
Comment (CasperWA) [inline tests/static/test_mongo_config.json]: By adding it explicitly to the `test_config.json` file (as `null`) and not adding it explicitly here, you're also testing that it works for these two cases (i.e., you'll be adding the test of "what if the option is not supplied - will it still correctly instantiate as `null`?").
```suggestion
```

**Review by CasperWA — CHANGES_REQUESTED** [review]
Looks good.

I have no comments or suggested changes on the actual code, but more on the added documentation.

File: aiida_optimade/routers/structures.py (hunk 7 above)
Comment (unkcpz) [reply aiida_optimade/routers/structures.py]: Thanks! I open https://github.com/aiidateam/aiida-optimade/pull/388 to use reorder importers pre-commit hook.

File: .github/workflows/ci_tests.yml (hunk 1 above)
Comment (unkcpz) [reply .github/workflows/ci_tests.yml]: Added, but somehow the coverage still decreases.

Comment (CasperWA) [reply .github/workflows/ci_tests.yml]: Could you test commenting out this step in the job and see if the coverage changes? I might be good to understand what the issue is. This _should_ work, meaning the coverage found through this step should be added to the existing `coverage.xml` file. It might however work if the coverage file name is added explicitly? I.e., changing the `--cov-report` value from `xml` to `xml:coverage.xml` for both this step and the step above with the first `pytest` run.

Comment (unkcpz) [reply .github/workflows/ci_tests.yml]: Sorry, I just saw this, I'll test this and request your review then.

Comment (CasperWA) [reply .github/workflows/ci_tests.yml]: Did you look into this?

Comment (unkcpz) [reply .github/workflows/ci_tests.yml]: Move the test right after the AiiDA test with `--cov-append` works.

**Comment by unkcpz** [comment]
@CasperWA All tests include the code coverage passed. Cheers! 🎉

File: README.md [hunk 9]
```diff
@@ -153,6 +153,12 @@ You can configure the server with the `aiida_optimade/config.json` file or set c
 
 To learn more about this, see the [`optimade-python-tools`](https://github.com/Materials-Consortia/optimade-python-tools) repository.
 
+### Using AiiDA group for curated data
+
+An [AiiDA Group](https://aiida.readthedocs.io/projects/aiida-core/en/latest/howto/data.html#organizing-data) can be used to curate data and server only this curated data through the OPTIMADE server.
```
Comment (CasperWA) [inline README.md]:
```suggestion
An [AiiDA Group](https://aiida.readthedocs.io/projects/aiida-core/en/latest/howto/data.html#organizing-data) can be used to curate data and serve only this curated data through the OPTIMADE server.
```

**Review by CasperWA — CHANGES_REQUESTED** [review]
Looking good now - only a minor typo change suggestion - then I'll appprove and merge 👍

File: README.md (hunk 9 above)
Comment (unkcpz) [reply README.md]: Ha, it was changed back after I did the rebase 😅 Done

**Review by CasperWA — APPROVED** [review]
This is an excellent addition to this server 💪
